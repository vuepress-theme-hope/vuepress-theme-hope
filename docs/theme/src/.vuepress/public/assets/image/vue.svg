<svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 128 128"><path fill="#42b883" d="M78.8 10 64 35.4 49.2 10H0l64 110 64-110z"/><path fill="#35495e" d="M78.8 10 64 35.4 49.2 10H25.6L64 76l38.4-66z"/></svg>