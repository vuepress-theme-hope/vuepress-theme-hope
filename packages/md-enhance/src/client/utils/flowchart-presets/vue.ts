import base from "./base.js";

export default {
  ...base,

  // style symbol types
  symbols: {
    start: {
      class: "start-element",
      "font-color": "#fff",
      fill: "#2F495F",
      "line-width": "0px",
    },
    end: {
      class: "end-element",
      "font-color": "#fff",
      fill: "#2F495F",
      "line-width": "0px",
    },
    operation: {
      class: "operation-element",
      "font-color": "#fff",
      fill: "#00BC7D",
      "line-width": "0px",
    },
    inputoutput: {
      class: "inputoutput-element",
      "font-color": "#fff",
      fill: "#EB4D5D",
      "line-width": "0px",
    },
    subroutine: {
      class: "subroutine-element",
      "font-color": "#fff",
      fill: "#937AC4",
      "element-color": "#fff",
      "line-color": "red",
    },
    condition: {
      class: "condition-element",
      "font-color": "#fff",
      fill: "#FFB500",
      "line-width": "0px",
    },
    parallel: {
      class: "parallel-element",
      "font-color": "#fff",
      fill: "#2F495F",
      "line-width": "0px",
    },
  },
};
