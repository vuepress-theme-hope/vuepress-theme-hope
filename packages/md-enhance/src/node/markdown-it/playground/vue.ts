import { deepAssign, entries, fromEntries } from "vuepress-shared/node";

import {
  type PlaygroundData,
  type PlaygroundOptions,
  type VuePresetPlaygroundOptions,
} from "../../typings/index.js";

const VUE_SUPPORTED_EXTENSIONS = [
  "html",
  "js",
  "ts",
  "vue",
  "jsx",
  "tsx",
  "json",
];

const DEFAULT_VUE_CDN = "https://sfc.vuejs.org/vue.runtime.esm-browser.js";
const DEFAULT_VUE_SR_CDN =
  "https://sfc.vuejs.org/server-renderer.esm-browser.js";

export const getVuePlaygroundPreset = (
  options: VuePresetPlaygroundOptions = {}
): PlaygroundOptions => ({
  name: "playground#vue",
  propsGetter: (playgroundData: PlaygroundData): Record<string, string> => {
    const { title = "", files, settings: localSettings, key } = playgroundData;
    const settings = {
      // defaults
      service: "https://sfc.vuejs.org/",
      dev: false,
      ssr: false,
      ...options,
      ...localSettings,
    };

    const fileInfo = fromEntries(
      entries(files)
        .filter(([, { ext }]) => VUE_SUPPORTED_EXTENSIONS.includes(ext))
        .map(([key, { content }]) => {
          if (key === "import-map.json") {
            const importMap = <
              {
                imports: Record<string, string>;
                scopes?: Record<string, Record<string, string>>;
              }
            >JSON.parse(content);

            return [
              key,
              JSON.stringify(
                deepAssign(
                  {
                    imports: {
                      // insure vue exists
                      vue: DEFAULT_VUE_CDN,
                      // insure vue/server-renderer exists
                      ...(settings.ssr
                        ? {
                            // eslint-disable-next-line @typescript-eslint/naming-convention
                            "vue/server-renderer": DEFAULT_VUE_SR_CDN,
                          }
                        : {}),
                    },
                  },
                  importMap
                )
              ),
            ];
          }

          return [key, content];
        })
    );

    if (settings.ssr && !fileInfo["import-map.json"])
      fileInfo["import-map.json"] = JSON.stringify(
        {
          imports: {
            vue: DEFAULT_VUE_CDN,
            // eslint-disable-next-line @typescript-eslint/naming-convention
            "vue/server-renderer": DEFAULT_VUE_SR_CDN,
          },
        },
        null,
        2
      );

    return {
      key,
      title,
      link: encodeURIComponent(
        `${settings.service}#${
          // dev flag
          settings.dev ? "__DEV__" : ""
        }${
          // ssr flag
          settings.ssr ? "__SSR__" : ""
        }${
          // code base64
          Buffer.from(JSON.stringify(fileInfo)).toString("base64")
        }`
      ),
    };
  },
});
