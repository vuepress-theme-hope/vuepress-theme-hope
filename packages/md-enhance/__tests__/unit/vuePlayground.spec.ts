import MarkdownIt from "markdown-it";
import { describe, expect, it } from "vitest";

import { vuePlayground } from "../../src/node/markdown-it/index.js";

describe("Vue Playground", () => {
  const markdownIt = MarkdownIt({ linkify: true }).use(vuePlayground);

  it("Should resolve playground info", () => {
    const internalResult = markdownIt.render(
      `
::: vue-playground Playground demo

@file App.vue
\`\`\`vue
<script setup>
import { ref } from 'vue'
const msg = ref('Hello World!')
</script>
<template>
  <h1>{{ msg }}</h1>
  <input v-model="msg" />
</template>
\`\`\`

@file Comp.vue

\`\`\`vue
<template>
  <div>Comp</div>
</template>
\`\`\`

@import

\`\`\`json
{
  "imports": {
    "vue": "https://sfc.vuejs.org/vue.runtime.esm-browser.js"
  }
}
\`\`\`
:::
`,
      {}
    );

    expect(internalResult).toMatchSnapshot();

    expect(
      markdownIt.render(
        `
::: vue-playground Playground demo

@file App.vue
\`\`\`vue
<script setup>
import { ref } from 'vue'
const msg = ref('Hello World!')
</script>
<template>
  <h1>{{ msg }}</h1>
  <input v-model="msg" />
</template>
\`\`\`

@file Comp.vue

\`\`\`vue
<template>
  <div>Comp</div>
</template>
\`\`\`

@import

\`\`\`json
{
  "imports": {
    "vue": "https://sfc.vuejs.org/vue.runtime.esm-browser.js"
  }
}
\`\`\`
:::
`,
        {}
      )
    ).toMatchSnapshot();
  });

  it("Should resolve playground info with settings", () => {
    expect(
      markdownIt.render(
        `
::: vue-playground Playground demo2

@file App.vue

\`\`\`vue
<script setup>
import { ref } from 'vue'
const msg = ref('Hello World!')
</script>
<template>
  <h1>{{ msg }}</h1>
  <input v-model="msg" />
</template>
\`\`\`

@file Comp.vue

\`\`\`vue
<template>
  <div>Comp</div>
</template>
\`\`\`

@setting

\`\`\`json
{
  "service": "https://element-plus.run/"
}
\`\`\`

:::
`,
        {}
      )
    ).toMatchSnapshot();

    expect(
      markdownIt.render(
        `
::: vue-playground Playground demo2

@file App.vue

\`\`\`vue
<script setup>
import { ref } from 'vue'
const msg = ref('Hello World!')
</script>
<template>
  <h1>{{ msg }}</h1>
  <input v-model="msg" />
</template>
\`\`\`

@file Comp.vue

\`\`\`vue
<template>
  <div>Comp</div>
</template>
\`\`\`

@setting

\`\`\`json
{
  "service": "https://element-plus.run/"
}
\`\`\`

:::
`,
        {}
      )
    ).toMatchSnapshot();
  });

  it("Should resolve playground info with settings", () => {
    expect(
      markdownIt.render(
        `
::: vue-playground Playground demo2

@file App.vue

\`\`\`vue
<script setup>
import { ref } from 'vue'
const msg = ref('Hello World!')
</script>
<template>
  <h1>{{ msg }}</h1>
  <input v-model="msg" />
</template>
\`\`\`

@import

\`\`\`json
{
  "imports": {
    "vue": "https://sfc.vuejs.org/vue.runtime.esm-browser.js"
  }
}
\`\`\`

@setting

\`\`\`json
{
  "service": "https://element-plus.run/"
}
\`\`\`

:::
`,
        {}
      )
    ).toMatchSnapshot();

    expect(
      markdownIt.render(
        `
::: vue-playground Playground demo2

@file App.vue

\`\`\`vue
<script setup>
import { ref } from 'vue'
const msg = ref('Hello World!')
</script>
<template>
  <h1>{{ msg }}</h1>
  <input v-model="msg" />
</template>
\`\`\`

@import

\`\`\`json
{
  "imports": {
    "vue": "https://sfc.vuejs.org/vue.runtime.esm-browser.js"
  }
}
\`\`\`

@setting

\`\`\`json
{
  "service": "https://element-plus.run/"
}
\`\`\`

:::
`,
        {}
      )
    ).toMatchSnapshot();
  });
});
