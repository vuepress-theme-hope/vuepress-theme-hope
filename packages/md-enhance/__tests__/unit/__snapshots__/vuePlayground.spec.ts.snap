// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Vue Playground > Should resolve playground info 1`] = `
"<VuePlayground title=\\"Playground%20demo\\" key=\\"62b3bf5a\\" settings=\\"%7B%7D\\" files=\\"eyJBcHAudnVlIjoiPHNjcmlwdCBzZXR1cD5cbmltcG9ydCB7IHJlZiB9IGZyb20gJ3Z1ZSdcbmNvbnN0IG1zZyA9IHJlZignSGVsbG8gV29ybGQhJylcbjwvc2NyaXB0PlxuPHRlbXBsYXRlPlxuICA8aDE%2Be3sgbXNnIH19PC9oMT5cbiAgPGlucHV0IHYtbW9kZWw9XCJtc2dcIiAvPlxuPC90ZW1wbGF0ZT5cbiIsIkNvbXAudnVlIjoiPHRlbXBsYXRlPlxuICA8ZGl2PkNvbXA8L2Rpdj5cbjwvdGVtcGxhdGU%2BXG4iLCJpbXBvcnQtbWFwLmpzb24iOiJ7XG4gIFwiaW1wb3J0c1wiOiB7XG4gICAgXCJ2dWVcIjogXCJodHRwczovL3NmYy52dWVqcy5vcmcvdnVlLnJ1bnRpbWUuZXNtLWJyb3dzZXIuanNcIlxuICB9XG59XG4ifQ%3D%3D\\">
</VuePlayground>
"
`;

exports[`Vue Playground > Should resolve playground info 2`] = `
"<VuePlayground title=\\"Playground%20demo\\" key=\\"62b3bf5a\\" settings=\\"%7B%7D\\" files=\\"eyJBcHAudnVlIjoiPHNjcmlwdCBzZXR1cD5cbmltcG9ydCB7IHJlZiB9IGZyb20gJ3Z1ZSdcbmNvbnN0IG1zZyA9IHJlZignSGVsbG8gV29ybGQhJylcbjwvc2NyaXB0PlxuPHRlbXBsYXRlPlxuICA8aDE%2Be3sgbXNnIH19PC9oMT5cbiAgPGlucHV0IHYtbW9kZWw9XCJtc2dcIiAvPlxuPC90ZW1wbGF0ZT5cbiIsIkNvbXAudnVlIjoiPHRlbXBsYXRlPlxuICA8ZGl2PkNvbXA8L2Rpdj5cbjwvdGVtcGxhdGU%2BXG4iLCJpbXBvcnQtbWFwLmpzb24iOiJ7XG4gIFwiaW1wb3J0c1wiOiB7XG4gICAgXCJ2dWVcIjogXCJodHRwczovL3NmYy52dWVqcy5vcmcvdnVlLnJ1bnRpbWUuZXNtLWJyb3dzZXIuanNcIlxuICB9XG59XG4ifQ%3D%3D\\">
</VuePlayground>
"
`;

exports[`Vue Playground > Should resolve playground info with settings 1`] = `
"<VuePlayground title=\\"Playground%20demo2\\" key=\\"1877a7d0\\" settings=\\"%7B%22service%22%3A%22https%3A%2F%2Felement-plus.run%2F%22%7D\\" files=\\"eyJBcHAudnVlIjoiPHNjcmlwdCBzZXR1cD5cbmltcG9ydCB7IHJlZiB9IGZyb20gJ3Z1ZSdcbmNvbnN0IG1zZyA9IHJlZignSGVsbG8gV29ybGQhJylcbjwvc2NyaXB0PlxuPHRlbXBsYXRlPlxuICA8aDE%2Be3sgbXNnIH19PC9oMT5cbiAgPGlucHV0IHYtbW9kZWw9XCJtc2dcIiAvPlxuPC90ZW1wbGF0ZT5cbiIsIkNvbXAudnVlIjoiPHRlbXBsYXRlPlxuICA8ZGl2PkNvbXA8L2Rpdj5cbjwvdGVtcGxhdGU%2BXG4ifQ%3D%3D\\">
</VuePlayground>
"
`;

exports[`Vue Playground > Should resolve playground info with settings 2`] = `
"<VuePlayground title=\\"Playground%20demo2\\" key=\\"1877a7d0\\" settings=\\"%7B%22service%22%3A%22https%3A%2F%2Felement-plus.run%2F%22%7D\\" files=\\"eyJBcHAudnVlIjoiPHNjcmlwdCBzZXR1cD5cbmltcG9ydCB7IHJlZiB9IGZyb20gJ3Z1ZSdcbmNvbnN0IG1zZyA9IHJlZignSGVsbG8gV29ybGQhJylcbjwvc2NyaXB0PlxuPHRlbXBsYXRlPlxuICA8aDE%2Be3sgbXNnIH19PC9oMT5cbiAgPGlucHV0IHYtbW9kZWw9XCJtc2dcIiAvPlxuPC90ZW1wbGF0ZT5cbiIsIkNvbXAudnVlIjoiPHRlbXBsYXRlPlxuICA8ZGl2PkNvbXA8L2Rpdj5cbjwvdGVtcGxhdGU%2BXG4ifQ%3D%3D\\">
</VuePlayground>
"
`;

exports[`Vue Playground > Should resolve playground info with settings 3`] = `
"<VuePlayground title=\\"Playground%20demo2\\" key=\\"1877a7d0\\" settings=\\"%7B%22service%22%3A%22https%3A%2F%2Felement-plus.run%2F%22%7D\\" files=\\"eyJBcHAudnVlIjoiPHNjcmlwdCBzZXR1cD5cbmltcG9ydCB7IHJlZiB9IGZyb20gJ3Z1ZSdcbmNvbnN0IG1zZyA9IHJlZignSGVsbG8gV29ybGQhJylcbjwvc2NyaXB0PlxuPHRlbXBsYXRlPlxuICA8aDE%2Be3sgbXNnIH19PC9oMT5cbiAgPGlucHV0IHYtbW9kZWw9XCJtc2dcIiAvPlxuPC90ZW1wbGF0ZT5cbiIsImltcG9ydC1tYXAuanNvbiI6IntcbiAgXCJpbXBvcnRzXCI6IHtcbiAgICBcInZ1ZVwiOiBcImh0dHBzOi8vc2ZjLnZ1ZWpzLm9yZy92dWUucnVudGltZS5lc20tYnJvd3Nlci5qc1wiXG4gIH1cbn1cbiJ9\\">
</VuePlayground>
"
`;

exports[`Vue Playground > Should resolve playground info with settings 4`] = `
"<VuePlayground title=\\"Playground%20demo2\\" key=\\"1877a7d0\\" settings=\\"%7B%22service%22%3A%22https%3A%2F%2Felement-plus.run%2F%22%7D\\" files=\\"eyJBcHAudnVlIjoiPHNjcmlwdCBzZXR1cD5cbmltcG9ydCB7IHJlZiB9IGZyb20gJ3Z1ZSdcbmNvbnN0IG1zZyA9IHJlZignSGVsbG8gV29ybGQhJylcbjwvc2NyaXB0PlxuPHRlbXBsYXRlPlxuICA8aDE%2Be3sgbXNnIH19PC9oMT5cbiAgPGlucHV0IHYtbW9kZWw9XCJtc2dcIiAvPlxuPC90ZW1wbGF0ZT5cbiIsImltcG9ydC1tYXAuanNvbiI6IntcbiAgXCJpbXBvcnRzXCI6IHtcbiAgICBcInZ1ZVwiOiBcImh0dHBzOi8vc2ZjLnZ1ZWpzLm9yZy92dWUucnVudGltZS5lc20tYnJvd3Nlci5qc1wiXG4gIH1cbn1cbiJ9\\">
</VuePlayground>
"
`;
